paddle/fluid/framework/fleet/box_wrapper.h

```diff
@@ -14,27 +14,117 @@ limitations under the License. */
 #include <glog/logging.h>
+#include <algorithm>
+#include <atomic>
+#include <ctime>
+#include <deque>
+#include <map>
 #include <memory>
 #include <mutex>  // NOLINT
 #include <string>
+#include <unordered_set>
 #include <vector>
 #include "paddle/fluid/framework/data_set.h"
-#ifdef PADDLE_WITH_BOX_PS
-#include <boxps.h>
-#endif
+#include "paddle/fluid/framework/lod_tensor.h"
 #include "paddle/fluid/platform/gpu_info.h"
 #include "paddle/fluid/platform/place.h"
+#include "paddle/fluid/platform/timer.h"
 
 namespace paddle {
 namespace framework {
 
+#ifdef PADDLE_WITH_BOX_PS
+class BasicAucCalculator {
+ public:
+  BasicAucCalculator() {}
+  void init(int table_size) { set_table_size(table_size); }
+  void reset() {
+    for (int i = 0; i < 2; i++) {
+      _table[i].assign(_table_size, 0.0);
+    }
+    _local_abserr = 0;
+    _local_sqrerr = 0;
+    _local_pred = 0;
+  }
+  void add_data(double pred, int label) {
+    PADDLE_ENFORCE_GE(pred, 0.0, platform::errors::PreconditionNotMet(
+                                     "pred should be greater than 0"));
+    PADDLE_ENFORCE_LE(pred, 1.0, platform::errors::PreconditionNotMet(
+                                     "pred should be lower than 1"));
+    PADDLE_ENFORCE_EQ(
+        label * label, label,
+        platform::errors::PreconditionNotMet(
+            "label must be equal to 0 or 1, but its value is: %d", label));
+    int pos = std::min(static_cast<int>(pred * _table_size), _table_size - 1);
+    PADDLE_ENFORCE_GE(
+        pos, 0,
+        platform::errors::PreconditionNotMet(
+            "pos must be equal or greater than 0, but its value is: %d", pos));
+    PADDLE_ENFORCE_LT(
+        pos, _table_size,
+        platform::errors::PreconditionNotMet(
+            "pos must be less than table_size, but its value is: %d", pos));
+    std::lock_guard<std::mutex> lock(_table_mutex);
+    _local_abserr += fabs(pred - label);
+    _local_sqrerr += (pred - label) * (pred - label);
+    _local_pred += pred;
+    _table[label][pos]++;
+  }
+  void compute();
+  int table_size() const { return _table_size; }
+  double bucket_error() const { return _bucket_error; }
+  double auc() const { return _auc; }
+  double mae() const { return _mae; }
+  double actual_ctr() const { return _actual_ctr; }
+  double predicted_ctr() const { return _predicted_ctr; }
+  double size() const { return _size; }
+  double rmse() const { return _rmse; }
+  std::vector<double>& get_negative() { return _table[0]; }
+  std::vector<double>& get_postive() { return _table[1]; }
+  double& local_abserr() { return _local_abserr; }
+  double& local_sqrerr() { return _local_sqrerr; }
+  double& local_pred() { return _local_pred; }
+  void calculate_bucket_error();
+
+ protected:
+  double _local_abserr = 0;
+  double _local_sqrerr = 0;
+  double _local_pred = 0;
+  double _auc = 0;
+  double _mae = 0;
+  double _rmse = 0;
+  double _actual_ctr = 0;
+  double _predicted_ctr = 0;
+  double _size;
+  double _bucket_error = 0;
```

Comment:
in order to distinguish the join-calculator and the update-calculator, it is suggested that these statistics stored in a LoDTensor carried by corresponding scope.

Reply:
Thanks, will make it a future work.